### PR TITLE
Null inferences do not flow out of a finally block.

### LINKED ISSUE #34018: Inferred nullable flow state from a finally block
I don't think our current inference design interacts with finally blocks well:
``` c#
C? c = null;
try
{
    c = SomeNonNullComputation;
}
finally
{
    if (c != null) c.Cleanup();
}
c.Operation(); // undeserved warning
```
We infer from `c != null` that `c` might be null.  That inference leaks out to the enclosing construct.  The result is a warning when `c` is used after the try-finally statement.  This will be a common pain point.

I don't think indirect inferences from inside a `finally` block should leak out to the enclosing context.  An inference from an actual assignment in the `finally` block should indeed leak out, though.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #34018

### CODE REVIEW

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a couple of reviews of this tiny nullable change?

**Comment by gafter** [comment]
@jcouv We don't have a spec for how nullable inference is supposed to deal with try statements.  I expect that is something that we will do later.

**Comment by jcouv** [comment]
Then I would recommend adding some notes to the design doc, so that we record that we differentiate between assignments and inferences, and that those affect try statements. 
Better docs might come later, but we can do something good enough now.

**Comment by gafter** [comment]
@jcouv I've added the documentation you suggested.  If all is good, please review/approve.

@dotnet/roslyn-compiler May I please have a second review of this tiny (but subtle) change?

File: docs/features/nullable-reference-types.md [hunk 1]
```diff
@@ -271,6 +271,9 @@ static void F<T>(T x, T y)
 The top-level nullability of `x ?? y` is `!` if `x` is `!` and otherwise the top-level nullability of `y`.
 A warning is reported if there is a nested nullability mismatch between `x` and `y`.
 
+### Try-finally
+We infer the state after a try statement in part by keeping track of which variables may have been assigned a null value in the finally block. This tracking distinguishes actual assignments from inferences: only actual assignments (but not inferences) affect the state after the try statement.  See https://github.com/dotnet/roslyn/issues/34018 and https://github.com/dotnet/roslyn/pull/35276.  This is not yet reflected in the language specification for nullable reference types (as we don't have a specification for how to handle try statements at this time).
```
Comment (333fred) [inline docs/features/nullable-reference-types.md]: Please put new sentences on new lines.

**Pull request merged**